src/modules/mc_pos_control/Takeoff/Takeoff.cpp

```diff
@@ -0,0 +1,127 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file Takeoff.cpp
+ */
+
+#include "Takeoff.hpp"
+#include <mathlib/mathlib.h>
+
+void Takeoff::generateInitialValue(const float hover_thrust, float velocity_p_gain)
+{
+	velocity_p_gain = math::max(velocity_p_gain, 0.01f);
+	_takeoff_ramp_vz_init = -hover_thrust / velocity_p_gain;
+}
+
+void Takeoff::updateTakeoffState(const bool armed, const bool landed, const bool want_takeoff,
+				 const float takeoff_desired_vz, const bool skip_takeoff)
+{
+	_spoolup_time_hysteresis.set_state_and_update(armed);
+
+	switch (_takeoff_state) {
+	case TakeoffState::disarmed:
+		if (armed) {
+			_takeoff_state = TakeoffState::spoolup;
+
+		} else {
+			break;
+		}
+
+	case TakeoffState::spoolup:
+		if (_spoolup_time_hysteresis.get_state()) {
+			_takeoff_state = TakeoffState::ready_for_takeoff;
+
+		} else {
+			break;
+		}
+
+	case TakeoffState::ready_for_takeoff:
+		if (want_takeoff) {
+			_takeoff_state = TakeoffState::rampup;
+			_takeoff_ramp_vz = _takeoff_ramp_vz_init;
+
+		} else {
+			break;
+		}
+
+	case TakeoffState::rampup:
+		if (_takeoff_ramp_vz >= takeoff_desired_vz) {
+			_takeoff_state = TakeoffState::flight;
+
+		} else {
+			break;
+		}
+
+	case TakeoffState::flight:
+		if (landed) {
+			_takeoff_state = TakeoffState::ready_for_takeoff;
+		}
+
+		break;
+
+	default:
+		break;
+	}
+
+	if (armed && skip_takeoff) {
+		_takeoff_state = TakeoffState::flight;
+	}
+
+	// TODO: need to consider free fall here
+	if (!armed) {
+		_takeoff_state = TakeoffState::disarmed;
+	}
+}
+
+float Takeoff::updateRamp(const float dt, const float takeoff_desired_vz)
```

Comment:
@MaEtUgR Let's add a description to this function:
- this function is responsible for ramping up a reference vertical velocity setpoint (_takeoff_ramp_vz) during takeoff
- the reference velocity at the beginning of a takeoff is initialized to a value which when passed to the velocity controller results in a zero throttle setpoint
- by ramping up _takeoff_ramp_vz during the takeoff and using it to constain the maximum climb rate demanded by the active flight task a smooth takeoff behavior is achieved

Reply:
I amended it a bit and also made a description for `generateInitialValue()` because that's where part of the description belongs.